src/dotty/tools/dotc/core/TypeApplications.scala

```diff
@@ -469,30 +462,52 @@ class TypeApplications(val self: Type) extends AnyVal {
         }
       case nil => t
     }
+    val stripped = self.stripTypeVar
+    val dealiased = stripped.safeDealias
     if (args.isEmpty || ctx.erasedTypes) self
-    else self.stripTypeVar.safeDealias match {
-      case self: TypeLambda =>
-        if (!args.exists(_.isInstanceOf[TypeBounds])) self.instantiate(args)
-        else {
-          val reducer = new Reducer(self, args)
-          val reduced = reducer(self.resType)
-          if (reducer.allReplaced) reduced
-          else HKApply(self, args)
-        }
-      case self: PolyType =>
-        self.instantiate(args)
-      case self: AndOrType =>
-        self.derivedAndOrType(self.tp1.appliedTo(args), self.tp2.appliedTo(args))
-      case self: TypeAlias =>
-        self.derivedTypeAlias(self.alias.appliedTo(args))
-      case self: TypeBounds =>
-        self.derivedTypeBounds(self.lo, self.hi.appliedTo(args))
-      case self: LazyRef =>
-        LazyRef(() => self.ref.appliedTo(args))
-      case self: WildcardType =>
-        self
-      case self: TypeRef if self.symbol == defn.NothingClass =>
-        self
+    else dealiased match {
+      case dealiased: TypeLambda =>
+        def tryReduce =
+          if (!args.exists(_.isInstanceOf[TypeBounds])) {
+            val reduced = dealiased.instantiate(args)
+            if (dealiased eq stripped) reduced
+            else reduced match {
+              case AppliedType(tycon, args) if variancesConform(typParams, tycon.typeParams) =>
+                // Reducing is safe for type inference, as kind of type constructor does not change
```

Comment:
Could we add a `Config` boolean to never reduce to check that reducing does not affect semantics?